api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java

```diff
@@ -740,7 +751,249 @@ public void shouldTestPurgeProgramAttributeType() throws Exception {
+		Concept concept = Context.getConceptService().getAllConcepts().get(0);
+		User user = Context.getUserService().getAllUsers().get(0);
+		program.setConcept(concept);
+		program.setDescription("test");
+		program.setName("programTest");
+		program.setCreator(user);
+		List<Program> list = pws.getAllPrograms();
+		String uuid = program.getUuid();
+		pws.retireProgram(program, "abc");
+		pws.unretireProgram(program);
+		Program program1 = dao.getProgramByUuid(uuid);
+		assertEquals(program, program1);
+	}
+	
+	@Rule
+	public ExpectedException expectedException = ExpectedException.none();
+	
+	@Test
+	public void savePatientProgram_shouldTestThrowPatientStateRequiresException() {
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage("'PatientProgram(id=1, patient=Patient#2, program=Program(id=1, concept=Concept #1738, " +
+											"workflows=[ProgramWorkflow(id=1), ProgramWorkflow(id=2)]))' failed to validate with reason: states: {0} is required for a patient state");
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		for (PatientState state : patientProgram.getStates()) {
+			state.setState(null);
+		}
+		pws.savePatientProgram(patientProgram);
+	}
+	
+	@Test
+	public void savePatientProgram_shouldTestSetPatientProgram() {
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		for (PatientState state : patientProgram.getStates()) {
+			state.setPatientProgram(null);
+		}
+		pws.savePatientProgram(patientProgram);
+		for (PatientState state : patientProgram.getStates()) {
+			assertEquals(state.getPatientProgram(), patientProgram);
+		}
+	}
+	
+	@Rule
+	public ExpectedException expectedException2 = ExpectedException.none();
+	
+	@Test
+	public void savePatientProgram_shouldThrowPatientProgramAlreadyAssignedException() {
+		expectedException2.expect(APIException.class);
+		expectedException2.expectMessage("This PatientProgram contains a ProgramWorkflowState whose parent is already assigned " +
+											 "to PatientProgram(id=2, patient=Patient#2, program=Program(id=2, concept=Concept #10, workflows=[]))");
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		PatientProgram patientProgram1 = pws.getPatientProgram(2);
+		for (PatientState state : patientProgram.getStates()) {
+			state.setPatientProgram(patientProgram1);
+		}
+		pws.savePatientProgram(patientProgram);
+	}
+	
+	@Test
+	public void savePatientProgram_shouldTestSetState() {
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		pws.voidPatientProgram(patientProgram, "test");
+		for (PatientState state : patientProgram.getStates()) {
+			state.setVoided(false);
+			state.setVoidReason(null);
+		}
+		pws.savePatientProgram(patientProgram);
+		for (PatientState state : patientProgram.getStates()) {
+			assertEquals(state.getVoided(), true);
+			assertEquals(state.getVoidReason(), "test");
+		}
+	}
+	
+	@Test
+	public void getPrograms_shouldTestGetProgramsIfCohortIsEmpty() {
+		int i = 0;
```

Comment:
what is this variable?

Reply:
I left it there by accident. I removed this